File: Radix Sort/RadixSort-Kotlin/RadixSort.kt
package com.exam.kotlin.codingproject

import java.util.*

/**
 * Created by msaycon on 09,Oct,2019
 */
class RadixSort {

    companion object {

        // A utility function to get maximum value in arr[]
        fun getMax(arr: IntArray, n: Int): Int {
            var mx = arr[0]
            for (i in 1 until n)
                if (arr[i] > mx)
                    mx = arr[i]
            return mx
        }

        // does counting sort
        fun countSort(arr: IntArray, n: Int, exp: Int) {
            val output = IntArray(n)
            var i: Int
            val count = IntArray(10)
            Arrays.fill(count, 0)
            i = 0
            while (i < n) {
                count[arr[i] / exp % 10]++
                i++
            }
            i = 1
            while (i < 10) {
                count[i] += count[i - 1]
                i++
            }
            i = n - 1
            while (i >= 0) {
                output[count[arr[i] / exp % 10] - 1] = arr[i]
                count[arr[i] / exp % 10]--
                i--
            }

            i = 0
            while (i < n) {
                arr[i] = output[i]
                i++
            }
        }

        // The main function to that sorts numbersToSort[] of size n using Radix Sort
        fun radixsort(numbersToSort: IntArray, n: Int) : IntArray {
            // Find the maximum number to know number of digits
            val m = getMax(numbersToSort, n)

            // Do counting sort for every digit. Note that instead
            // of passing digit number, exp is passed. exp is 10^i
            // where i is current digit number
            var exp = 1
            while (m / exp > 0) {
                countSort(numbersToSort, n, exp)
                exp *= 10
            }

            return numbersToSort
        }
    }
}

File: Radix Sort/RadixSort-C/radix_Sort.c
// Radix sort C implementation
#include<stdio.h>
#include<conio.h>
// The main function to that sorts arr[] of size n using Radix Sort
void radix(int a[],int n)
{

    int max=a[0];
    for(int i=1;i<n;i++)
    {
        if(max<a[i])
        {
            max=a[i];
        }
    }
    // Do counting sort for every digit. Note that instead
    // of passing digit number, exp is passed. exp is 10^i
    // where i is current digit number
    for(int i=1;max/i>0;i*=10)
    {
        count(a,n,i);
    }
}
// does counting sort
void count(int a[],int n,int p)
{
    int max=a[0];
    for(int i=1;i<n;i++)
    {
        if(max<a[i])
        {
            max=a[i];
        }
    }
    int b[max+1];
    for(int i=0;i<max+1;i++)
    {
        b[i]=0;
    }
    for(int i=0;i<n;i++)
    {
        b[(a[i]/p)%10]+=1;
    }
    for(int i=1;i<max+1;i++)
    {
       b[i]+=b[i-1];
    }
    int c[n];
    for(int i=0;i<n;i++)
    {
        c[i]=0;
    }
    for(int i=0;i<n;i++)
    {
        c[b[(a[i]/p)%10]-1]=a[i];
        b[(a[i]/p)%10]-=1;
    }
    for(int i=0;i<n;i++)
    {
        a[i]=c[i];
    }
}
/*Driver function to check for above function*/
void main()
{
    int n;
    scanf("%d",&n);
    int arr[n];
    for(int i=0;i<n;i++)
    {
        scanf("%d",&arr[i]);
    }
    radix(arr,n);
    for(int i=0;i<n;i++)
    {
        printf("%d\n",arr[i]);
    }


}
